packages/react-storage/src/components/StorageBrowser/views/LocationActionView/CopyView/useDestinationPicker.ts

```diff
@@ -0,0 +1,124 @@
+import { useCallback, useEffect, useRef } from 'react';
+import { isString } from '@aws-amplify/ui';
+import { usePaginate } from '../../hooks/usePaginate';
+import {
+  listLocationItemsHandler,
+  ListLocationItemsHandlerInput,
+  ListLocationItemsHandlerOutput,
+} from '../../../actions/handlers/listLocationItems';
+import { useGetActionInput } from '../../../providers/configuration';
+import { getDestinationListFullPrefix } from '../utils/getDestinationPickerDataTable';
+
+import { useDataState } from '@aws-amplify/ui-react-core';
+import { useStore } from '../../../providers/store';
+
+const DEFAULT_PAGE_SIZE = 1000;
+export const DEFAULT_LIST_OPTIONS = {
+  pageSize: DEFAULT_PAGE_SIZE,
+  delimiter: '/',
+};
+
+const DEFAULT_REFRESH_OPTIONS = { ...DEFAULT_LIST_OPTIONS, refresh: true };
+
+export const useDestinationPicker = ({
+  destinationList,
+}: {
+  destinationList: string[];
+}): {
+  bucket: string | undefined;
+  items: ListLocationItemsHandlerOutput['items'];
+  hasNextToken: boolean;
+  currentPage: number;
+  isLoading: boolean;
+  hasError: boolean;
+  message: string | undefined;
+  handleNext: () => void;
+  handlePrevious: () => void;
+  range: [number, number];
+} => {
+  const previousPathref = useRef<string | null>(null);
+
+  const prefix = getDestinationListFullPrefix(destinationList);
+
+  const locationItemsAction = useCallback(
+    async (
+      previous: ListLocationItemsHandlerOutput,
+      input: ListLocationItemsHandlerInput
+    ): Promise<ListLocationItemsHandlerOutput> => {
+      const { items, nextToken } = await listLocationItemsHandler(input);
+      const newItems =
+        previousPathref.current !== input.prefix
+          ? items
+          : previous.items.concat(items);
+
+      previousPathref.current = prefix;
```

Comment:
Can we add a note explaining this logic in a follow up?

Reply:
When you navigate to a new folder the prefix will change and thus we should throw away the old items and show the new folders at the new prefix. The best way I could find to tell if you had navigated was to store the previous prefix in a ref and compare them.